### PR TITLE
Low level bulk operations

### PR DESCRIPTION
**Pull request by haiodo**
Signed-off-by: Andrey Sobolev <haiodo@gmail.com>

# Contribution checklist

## Brief description

* Rework server storage and adapters to support bulk operations.
* Fix deletion of attached documents + performance speedup.

## Checklist

* [x] - UI test added to added/changed functionality?
* [x] - Screenshot is added to PR if applicable ?
* [x] - Does the code work? Check whether function and logic are correct.
* [x] - Does Changelog.md is updated with changes?
* [x] - Does the translations are up to date?
* [x] - Does it well tested?
* [x] - Tested for Chrome.
* [x] - Tested for Safari.
* [x] - Go through the changed code looking for typos, TODOs, commented LOCs, debugging pieces of code, etc.
* [x] - Rebase your branch onto master and upstream branch
* [x] - Is there any redundant or duplicate code?
* [x] - Are required links are linked to PR?
* [x] - Does new code is well documented ?

## Related issues

A list of closed updated issues

### CODE REVIEW

File: plugins/view-resources/src/components/EditDoc.svelte [hunk 1]
```diff
@@ -70,7 +70,9 @@
   $: if (_id && _class) {
     query.query(_class, { _id }, (result) => {
       object = result[0]
-      realObjectClass = object._class
+      if (object !== undefined) {
+        realObjectClass = object?._class
```
Comment (BykhovDenis) [inline plugins/view-resources/src/components/EditDoc.svelte]: `if` or `?` is extra

File: dev/storage/src/storage.ts [hunk 2]
```diff
@@ -44,8 +44,15 @@ class InMemoryTxAdapter extends DummyDbAdapter implements TxAdapter {
...
 
-  tx (tx: Tx): Promise<TxResult> {
-    return this.txdb.tx(tx)
+  async tx (...tx: Tx[]): Promise<TxResult> {
+    const r: TxResult[] = []
+    for (const t of tx) {
+      r.push(await this.txdb.tx(t))
+    }
+    if (r.length === 0) {
+      return r[0]
```
Comment (BykhovDenis) [inline dev/storage/src/storage.ts:53]: Undefined? Probably it should be `if (r.length === 1)`, also function return type should be Promise<TxResult | TxResult[]>

File: server/core/src/storage.ts [hunk 3]
```diff
@@ -138,16 +139,50 @@ class TServerStorage implements ServerStorage {
...
+        // Skip unsupported tx
+        console.error('Unsupported transacton', tx)
+        continue
+      }
       const domain = this.hierarchy.getDomain(txCUD.objectClass)
-      const adapter = this.getAdapter(domain)
-      const res = await adapter.tx(txCUD)
-      return res
-    } else {
+      if (part.length > 0) {
```
Comment (BykhovDenis) [inline server/core/src/storage.ts:164]:
```suggestion
     if (lastDomain !== domain && part.length > 0) {
await processPart()
}
lastDomain = domain
part.push(txCUD)
```

File: server/core/src/storage.ts [hunk 4]
```diff
@@ -299,73 +342,140 @@ class TServerStorage implements ServerStorage {
...
+          { sort: { modifiedOn: 1 } }
+        )
+        : []
+    const result: Doc[] = []
+
+    const txesAttach =
+      removeAttachObjectIds.length > 0
+        ? await findAll<TxCollectionCUD<Doc, AttachedDoc>>(
+          ctx,
+          core.class.TxCollectionCUD,
```
Comment (BykhovDenis) [inline server/core/src/storage.ts:377]: I'm not sure that all AttachedDoc CUD ties were used CollectionTx

File: server/core/src/storage.ts [hunk 5]
```diff
@@ -299,73 +342,140 @@ class TServerStorage implements ServerStorage {
...
+        continue
+      }
+      const rtx = actualTx as TxRemoveDoc<Doc>
+      const isAttached = this.hierarchy.isDerived(rtx.objectClass, core.class.AttachedDoc)
+
+      const objTxex = isAttached
+        ? txesAttach.filter((tx) => tx.tx.objectId === rtx.objectId)
+        : txes.filter((it) => it.objectId === rtx.objectId)
+
+      const createTx = isAttached
```
Comment (BykhovDenis) [inline server/core/src/storage.ts]: We have `TxProcessor.buildDoc2Doc` for next 15 lines

File: server/core/src/storage.ts [hunk 6]
```diff
@@ -299,73 +342,140 @@ class TServerStorage implements ServerStorage {
...
+        ? await findAll<TxCollectionCUD<Doc, AttachedDoc>>(
+          ctx,
+          core.class.TxCollectionCUD,
+          { 'tx.objectId': { $in: removeAttachObjectIds } },
+          { sort: { modifiedOn: 1 } }
+        )
+        : []
+
+    for (const tx of rawTxes) {
+      const actualTx = TxProcessor.extractTx(tx)
```
Comment (BykhovDenis) [inline server/core/src/storage.ts]: We do it twice, I think rawTxes not big array and it's ok, but probably better create new array for store rtx and isAttached in first loop and use this array for second loop

File: server/core/src/storage.ts [hunk 7]
```diff
@@ -395,39 +512,48 @@ class TServerStorage implements ServerStorage {
...
+    for (const tx of txes) {
+      const actualTx = TxProcessor.extractTx(tx)
+      if (!this.hierarchy.isDerived(actualTx._class, core.class.TxUpdateDoc)) {
+        continue
+      }
+      const rtx = actualTx as TxUpdateDoc<Doc>
+      if (rtx.operations.space === undefined || rtx.operations.space === rtx.objectSpace) {
+        continue
+      }
+      const factory = new TxFactory(core.account.System)
```
Comment (BykhovDenis) [inline server/core/src/storage.ts]: I think it should be `tx.modifiedBy`

**Review by BykhovDenis — APPROVED** [review]

File: server/core/src/storage.ts (hunk 4 above)
Comment (haiodo) [reply server/core/src/storage.ts:377]: They should be, before it was same logic. So better they to be :-)

File: server/core/src/storage.ts (hunk 5 above)
Comment (haiodo) [reply server/core/src/storage.ts]: Thanks.

File: server/core/src/storage.ts (hunk 6 above)
Comment (haiodo) [reply server/core/src/storage.ts]: Fixed.

File: server/core/src/storage.ts (hunk 7 above)
Comment (haiodo) [reply server/core/src/storage.ts]: Fixed.

File: server/core/src/storage.ts (hunk 3 above)
Comment (haiodo) [reply server/core/src/storage.ts:164]: I do no think this check is really required, in place there it is needed it checked.

**Pull request merged**